algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,316 @@
+  }
+end
+
+@[simp] lemma mod_one (a : α) : a % 1 = 0 :=
+begin
+  cases val_mod a 1,
+  { 
+    have := one_ne_zero, 
+    contradiction,
+  },
+  {
+    exact val_lt_one (a % 1) h,
+  }
+end 
+
+@[simp] lemma zero_mod (b : α) : 0 % b = 0 :=
+begin
+  have h1 := witness 0 b,
+  have h2 : remainder 0 b = b * (-quotient 0 b ), from
+    calc
+      remainder 0 b = quotient 0 b * b + remainder 0 b + b * (-quotient 0 b ) : by ring
+      ...                       = b * (-quotient 0 b ) : by rw [h1, zero_add],
+  cases val_mul b (-quotient 0 b),
+  {
+    simp at h, rw h at h1, simp at h1,
+    exact h1,
+  },
+  {
+    rw [←h2] at h,
+    cases val_mod 0 b,
+    {
+      rw h_1, exact mod_zero (0:α),
+    },
+    {
+      have := not_le_of_lt h_1,
+      contradiction,
+    }
+  }
+end
+
+@[simp] lemma zero_div (b : α) (hnb : b ≠ 0) : 0 / b = 0 :=
+begin
+  have h1 : remainder 0 b = 0, from zero_mod b,
+  have h2 := witness 0 b,
+  rw h1 at h2,
+  simp at h2,
+  cases eq_zero_or_eq_zero_of_mul_eq_zero h2,
+  { exact h },
+  { contradiction }
+end
+
+@[simp] lemma mod_self (a : α) : a % a = 0 :=
+begin
+  have := witness a a,
+  cases (dvd_mod_self a) with m a_mul,
+  cases val_mul a m,
+  {
+    rw [h, mul_zero] at a_mul,
+    exact a_mul,
+  },
+  {
+    cases  val_mod a a, 
+    {rw h_1, exact mod_zero (0:α)},
+    {
+      have := not_le_of_lt h_1,
+      dsimp [(%)] at a_mul,
+      rw a_mul at this,
+      contradiction,
+    }
+  }
+end 
+
+lemma div_self (a : α) : a ≠ 0 → a / a = (1:α) :=
+begin
+  intro hna,
+  have wit_aa := witness a a,
+  have a_mod_a := mod_self a, 
+  dsimp [(%)] at a_mod_a,
+  rw a_mod_a at wit_aa, simp at wit_aa,
+  have h1 : 1 * a = a, from one_mul a,
+  conv at wit_aa {for a [4] {rw ←h1}},
+  exact eq_of_mul_eq_mul_right hna wit_aa
+end
+
+/- weak gcd lemmas -/
+
+@[simp] theorem gcd_zero_left (a : α) : gcd 0 a = a := 
+begin
+  rw gcd,
+  simp,
+end
+
+@[elab_as_eliminator]
+theorem gcd.induction {P : α → α → Prop}
+                  (a b : α)
+                  (H0 : ∀ x, P 0 x)
+                  (H1 : ∀ a b, a ≠ 0 → P (b%a) a → P a b) :
+                P a b := 
+@well_founded.induction _ _ (has_well_founded.wf α) (λa, ∀b, P a b) a (λc IH,
+  by {cases decidable.em (c=0), rw h, exact H0,
+    exact λ b, H1 c b (h) (IH (b%c) (neq_zero_lt_mod_lt b c h) c)}) b
+
+theorem gcd_dvd (a b : α) : (gcd a b ∣ a) ∧ (gcd a b ∣ b) :=
+gcd.induction a b
+  (λ b, by simp)
+  (λ a b aneq,
+  begin
+    intro h_dvd,
+    rw gcd, simp [aneq],
+    induction h_dvd,
+    split,
+    { exact h_dvd_right },
+    {
+      conv {for b [2] {rw ←(witness b a)}},
+      have h_dvd_right_a:= dvd_mul_of_dvd_right h_dvd_right (b/a),
+      exact dvd_add h_dvd_right_a h_dvd_left
+    }
+  end )
+
+theorem gcd_dvd_left (a b : α) : gcd a b ∣ a := (gcd_dvd a b).left
+
+theorem gcd_dvd_right (a b : α) : gcd a b ∣ b := (gcd_dvd a b).right
+
+theorem dvd_gcd {a b c : α} : c ∣ a → c ∣ b → c ∣ gcd a b :=
+gcd.induction a b
+  (λ b,
+  begin
```

Comment:
why `λ b,` followed by a `intros`? Just remove the parenthesis and move the lambda to the intros.